### PR TITLE
Add PyTorchEspresso estimator to support Espresso models for ASR

### PR DESCRIPTION
**Pull request by YiwenShaoStephen**
# Description

This module implements the task specific estimator for Espresso, an end-to-end speech recognition toolkit based on fairseq

## TODO list
- [ ] `fit` function
- [ ]  Proper ways to load Espresso config/models
- [ ]  Package dependency related changes
- [ ]  Test

### CODE REVIEW

**Comment by YiwenShaoStephen** [comment]
For now it's just a draft PR that only contains the minimal codes to make Espresso models work under ART. 
More issues need to be discussed later on here.
@beat-buesser Please don't hesitate to make any comments here to let me know what to add/change in this PR.

**Comment by beat-buesser** [comment]
Hi @YiwenShaoStephen

Thank you very much, this is great! I'll take closer look and add ideas very soon.

I'm also adding @minhitbk who has implemented `PyTorchDeepSpeech` for help and advice.

File: art/estimators/speech_recognition/pytorch_espresso.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        from argparse import Namespace
+        from fairseq import checkpoint_utils, tasks, utils
+        from fairseq.data import encoders
+        import sentencepiece as spm
+
+        # Super initialization
+        super().__init__(
+            model=None,
+            clip_values=clip_values,
+            channels_first=None
```
Comment (beat-buesser) [inline art/estimators/speech_recognition/pytorch_espresso.py:90]: @YiwenShaoStephen I think this missing comma makes the CI cail:
```suggestion
            channels_first=None,
```

**Comment by beat-buesser** [comment]
@YiwenShaoStephen Does `PyTorchEspresso` have additional dependencies beyond what can be installed with `pip`? If yes, do you have a `Dockerfile` that defines an envrionment for Espresso? @minhitbk has very recently created a Dockerfile for `PyTorchDeepSpeech` that we now use in a new GitHub Actions workflow for unit testing `PyTorchDeepSpeech`: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/d18fbd32730d864aff6ea3316e68b42ab08a385b/.github/actions/deepspeech-v2/Dockerfile
Maybe this could be helpful for unit testing `PyTorchEspresso`.

**Comment by minhitbk** [comment]
@YiwenShaoStephen: Is this version of the EspressoEstimator can work directly with the current implementation of the imperceptible_asr_attack_pytorch without modifying/changing the attack? In your previous PR (https://github.com/YiwenShaoStephen/adversarial-robustness-toolbox/pull/1/files), it seems that you need to change the attack?

**Comment by pzelasko** [comment]
BTW once the code is in good shape, we can also share our pre-trained LibriSpeech models if you have the option to host them somewhere.

**Comment by beat-buesser** [comment]
Hi @pzelasko That sounds great! So far ART only has a very small storage for a few, small, pretrained models that are used in some of the ART notebooks. But several of the model-specific `art.estimators` automatically download the model weights from external model zoos where available.

**Comment by YiwenShaoStephen** [comment]
@minhitbk Probably not. As is written in the imperceptible_asr_attack_pytorch (https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/dev_1.7.0/art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py#L48),  it is hard coded to work with `PyTorchDeepSpeech` only. It would be great if some one can make this attack more flexible to support all `PyTorchEstimator`, otherwise an extra attack might be necessary to added to support `EspressoEstimator`.

**Comment by YiwenShaoStephen** [comment]
@beat-buesser With the latest update of Espresso, it can be `pip` install all the way now.

**Comment by beat-buesser** [comment]
> @minhitbk Probably not. As is written in the imperceptible_asr_attack_pytorch (https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/dev_1.7.0/art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py#L48), it is hard coded to work with `PyTorchDeepSpeech` only. It would be great if some one can make this attack more flexible to support all `PyTorchEstimator`, otherwise an extra attack might be necessary to added to support `EspressoEstimator`.

@YiwenShaoStephen Does @minhitbk 's #1052 address this issue?

**Review by beat-buesser — APPROVED** [review]
Hi @YiwenShaoStephen and @pzelasko Thank you very much for contributing the new `PyTorchEspresso` estimator for Espresso ASR and making it compatible with ART's FGM/PGD and Imperceptible ASR attacks! This establishes a great connection between Espresso and ART enabling research on adversarial machine learning in ASR!

**Comment by pzelasko** [comment]
Great! Thanks for your help in coordinating this PR 🙂

**Pull request merged**